oras/auth/token.py

```diff
@@ -73,17 +77,20 @@ def authenticate_request(
         h = auth_utils.parse_auth_header(authHeaderRaw)
 
         # First try to request an anonymous token
-        logger.debug("No Authorization, requesting anonymous token")
-        anon_token = self.request_anonymous_token(h)
-        if anon_token:
-            logger.debug("Successfully obtained anonymous token!")
-            self.token = anon_token
-            headers["Authorization"] = "Bearer %s" % self.token
-            return headers, True
+        if not skipAnonToken:
```

Comment:
Isn't there a more programmatic way to figure this out than provide a boolean? If they just don't provide any credentials can't we assume using anon?

Reply:
that it's a nice idea, could experiment with it, based on feedback from: https://github.com/oras-project/oras-py/pull/148#discussion_r1667623649

Reply:
> If they just don't provide any credentials can't we assume using anon?

I would like to double-check I'm understanding your suggestion correctly, please.

In the current main branch code (unrelated to this PR), the anon token is always tried first, and as far as I can tell, there is no way when the second call to `token.py authenticate_request()` happens to  check whether the anon token was tried, it actually always re-tries the anon token:

https://github.com/oras-project/oras-py/blob/caf8db5b279382335fbb1f6d7402ed9b73618d37/oras/auth/token.py#L46-L85

See line 77.
It appears to me, line 77 is always executed for asking a `anon_token = self.request_anonymous_token(h)  `
and there is no way to jump down to line 85, to request the auth token.

This is what appears to me is happening today on main branch.

With this PR, I've introduced a deliberate flag whether anon_token should be skipped (so avoid line 77, and jump to 85) and the 3rd attempt from `provider.py do_request()`.

What your suggestion now makes me wonder now, is whether to actually simplify the logic differently here as:
- anon token request is only tried if there is NO auth entry
- in case an auth entry is available, never try anon token, and just perform auth token

Can you kindly feedback @vsoch about this, please?
Thank you, and hope this helps.

Reply:
> it actually always re-tries the anon token

For request token, the basic auth is needed right? So perhaps that's the logic that should be added before showing the user:

```python
logger.debug("No Authorization, requesting anonymous token") 
```
And if the basic auth is there, skip over asking for an anon token. If there are cases where we should try the basic auth first and fall back to anonymous when needed, that could be done too.

Reply:
implemented with https://github.com/oras-project/oras-py/pull/148/commits/bb4c2ec85f8614dc9a21e3c9ea7d9feb388ff91c